logbus/logstreamer/logstreamer_orchestrator.go

```diff
@@ -0,0 +1,170 @@
+package logstreamer
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/earthly/cloud-api/logstream"
+	"github.com/hashicorp/go-multierror"
+)
+
+type LogstreamOrchestrator struct {
+	buildID         string
+	bus             LogBus
+	c               CloudClient
+	initialManifest *logstream.RunManifest
+
+	doneMU sync.Mutex
+	mu     sync.Mutex
+	doneCH chan struct{}
+
+	errMu       sync.Mutex
+	errors      []error
+	started     atomic.Bool
+	retries     int
+	streamer    *LogStreamer
+	deltas      *deltasIter
+	verbose     bool
+	deltaBuffer int
+}
+
+type LOpt func(*LogstreamOrchestrator) *LogstreamOrchestrator
+
+// WithVerbose sets the verbose option on the LogstreamOrchestrator
+func WithVerbose(verbose bool) func(orchestrator *LogstreamOrchestrator) *LogstreamOrchestrator {
+	return func(orchestrator *LogstreamOrchestrator) *LogstreamOrchestrator {
+		orchestrator.verbose = verbose
+		return orchestrator
+	}
+}
+
+func WithDeltaBuffer(buffer int) func(orchestrator *LogstreamOrchestrator) *LogstreamOrchestrator {
+	return func(orchestrator *LogstreamOrchestrator) *LogstreamOrchestrator {
+		orchestrator.deltaBuffer = buffer
+		return orchestrator
+	}
+}
+
+func NewLogstreamOrchestrator(bus LogBus, c CloudClient, initialManifest *logstream.RunManifest, opts ...LOpt) *LogstreamOrchestrator {
+	ls := &LogstreamOrchestrator{
+		buildID:         initialManifest.GetBuildId(),
+		bus:             bus,
+		c:               c,
+		initialManifest: initialManifest,
+		retries:         10,
+		deltaBuffer:     DefaultBufferSize,
+	}
+	for _, o := range opts {
+		ls = o(ls)
+	}
+	return ls
+}
+
+// StartLogstreamer will start streaming to the cloud retrying up the retry count
+// Callers should listen to Done to be notified when the streaming contract completes
+// StartLogstreamer may only be called once
+func (l *LogstreamOrchestrator) StartLogstreamer(ctx context.Context) {
+	if l.started.Swap(true) {
+		// Can only start once
+		return
+	}
+	go func() {
+		for i := 0; i < l.retries; i++ {
+			l.start()
+			l.CloseLastLogstreamer()
+
+			l.mu.Lock()
+			l.deltas = newDeltasIter(l.deltaBuffer, l.initialManifest, l.verbose)
+			l.streamer = New(l.c, l.buildID, l.deltas)
+			l.mu.Unlock()
+
+			go l.bus.AddSubscriber(l.deltas)
```

Comment:
It's important to note that doing this in a separate goroutine means that the subscriber _might not be added_ before it is removed. Since `RemoveSubscriber` doesn't check for that failure case (i.e. if the subscriber didn't exist, it happily returns as if the subscriber was removed), this can leave us with leaked subscribers.

We probably need some kind of communication for this - either a channel or a mutex lock:

```go
l.subCh = make(chan struct{})
go func() {
    defer close(l.subCh)
    l.bus.AddSubscriber(l.deltas)
}()
```

And in `l.CloseLastLogstreamer()`:
```go
if l.deltas != nil {
    <-l.subCh
    l.bus.RemoveSubscriber(l.deltas)
    l.deltas.close()
}
```
NOTE: this would need to be protected by `l.mu` as well, since it is reallocating the channel on each loop. Maybe the "separate method" I mentioned should be named `l.subscribe()` instead of `l.reset()` and take care of the `subCh` too.

Reply:
I've handled this in a `subscribe` function now.
Thanks for the feedback + good catch